pkg/recovery/recover.go

```diff
@@ -0,0 +1,370 @@
+// start` to re-bootstrap the control plane.
+func Recover(ctx context.Context, backend Backend, kubeConfigPath string) (asset.Assets, error) {
+	cp, err := backend.read(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	as, err := cp.renderSelfHosted()
+	if err != nil {
+		return nil, err
+	}
+
+	bs, err := cp.renderBootstrap(kubeConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	as = append(as, bs...)
+
+	ks, err := renderKubeConfig(kubeConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	as = append(as, ks...)
+
+	return as, nil
+}
+
+// renderSelfHosted returns assets for the self-hosted control plane objects. These are output
+// without modification from what the backend recovered.
+func (cp *controlPlane) renderSelfHosted() (asset.Assets, error) {
+	var as asset.Assets
+	if objAs, err := serializeListObjToYAML(&cp.configMaps); err != nil {
+		return nil, err
+	} else {
+		as = append(as, objAs...)
+	}
+	if objAs, err := serializeListObjToYAML(&cp.daemonSets); err != nil {
+		return nil, err
+	} else {
+		as = append(as, objAs...)
+	}
+	if objAs, err := serializeListObjToYAML(&cp.deployments); err != nil {
+		return nil, err
+	} else {
+		as = append(as, objAs...)
+	}
+	if objAs, err := serializeListObjToYAML(&cp.podDisruptionBudgets); err != nil {
+		return nil, err
+	} else {
+		as = append(as, objAs...)
+	}
+	if objAs, err := serializeListObjToYAML(&cp.secrets); err != nil {
+		return nil, err
+	} else {
+		as = append(as, objAs...)
+	}
+	if objAs, err := serializeListObjToYAML(&cp.services); err != nil {
+		return nil, err
+	} else {
+		as = append(as, objAs...)
+	}
+	if objAs, err := serializeListObjToYAML(&cp.serviceAccounts); err != nil {
+		return nil, err
+	} else {
+		as = append(as, objAs...)
+	}
+	return as, nil
+}
+
+// renderBootstrap returns assets for a bootstrap control plane that can be used with `bootkube
+// start` to re-bootstrap a control plane. These assets are derived from the self-hosted control
+// plane that was recovered by the backend, but modified for direct injection into a kubelet.
+func (cp *controlPlane) renderBootstrap(kubeConfigPath string) (asset.Assets, error) {
+	pods, err := extractBootstrapPods(cp.daemonSets.Items, cp.deployments.Items)
+	if err != nil {
+		return nil, err
+	}
+	requiredSecrets, err := fixUpBootstrapPods(pods)
+	if err != nil {
+		return nil, err
+	}
+	as, err := outputBootstrapPods(pods)
+	if err != nil {
+		return nil, err
+	}
+	secrets, err := outputBootstrapSecrets(cp.secrets.Items, requiredSecrets)
+	if err != nil {
+		return nil, err
+	}
+	as = append(as, secrets...)
+	return as, nil
+}
+
+// extractBootstrapPods extracts bootstrap pod specs from daemonsets and deployments.
+func extractBootstrapPods(daemonSets []v1beta1.DaemonSet, deployments []v1beta1.Deployment) ([]v1.Pod, error) {
+	var pods []v1.Pod
+	for _, ds := range daemonSets {
+		if componentName := ds.Labels["component"]; bootstrapComponents[componentName] {
```

Comment:
I'd say configurable might not be great, since we rely on other "bootkube conventions" such as various values we use for those labels.

Instead I added support for both the old and new label key, extracted to a function. Thanks for the nice catch.